Integration with `mcan` crate

glaeqen opened the pull request:
This commit provides `mcan` crate integration code

Tested with SAMV71 Xplained Ultra

glaeqen committed:
Integration with `mcan` crate

This commit provides `mcan` crate integration code

michalfita commented:
May I ask why [`mcan`](https://crates.io/crates/mcan-core) instead of [`embedded-can`](https://crates.io/crates/embedded-can)?

Wouldn't it be wise to gate this behind feature and make the dependency optional? CAN common in automotive industry, but rarely elsewhere.

Dzięki za wkład 👍🏼

michalfita commented:
It would be nice to add some usage example to board examples.

glaeqen commented:
> May I ask why [`mcan`](https://crates.io/crates/mcan-core) instead of [`embedded-can`](https://crates.io/crates/embedded-can)?

They are different things. `mcan-core` is a small crate that provides two traits that should be implemented within a target HAL in order to make `mcan` crate operational. `mcan` is an MCAN HAL that provides actual APIs to receive and send messages.

MCAN is the [Bosch IP](https://www.bosch-semiconductors.com/ip-modules/can-ip-modules/m-can/) that is integrated almost as-is by vendors into their MCUs, it has the same memory-mapped HW registers and same internals in every single MCU it has been integrated into.

User can link against `mcan` crate in their application crate if wants to and use `hal::can::Dependencies` to resolve platform-specific shenanigans in order to get it to work.

Also, we will be able to do breaking changes in `mcan` without a necessity of bumping and releasing target HALs that integrate with `mcan`. That's because `mcan-core` is supposed to be very stable. So `mcan` and target HALs can move independently.

`embedded-can` is just like `embedded-hal`, it allows driver implementors to have a generalized I2C or whatever so their library does not have to depend on any specific HAL. Drivers for devices that communicate over the I2C or CAN bus. 
In comparison, MCAN *is* the CAN peripheral embedded in the MCU.

> Wouldn't it be wise to gate this behind feature and make the dependency optional? CAN common in automotive industry, but rarely elsewhere.

This is a crate that does not contain any code. Two traits with big doc-strings. Code in a x7x HAL will be optimised away if not used in a final binary. Introduction of feature-gating seems like an unnecessary over-complication. To be fair, I did this implementation as an exercise, as other @GrepitAB folk uses V71 so I had development boards within a reach and I wanted to verify `mcan` assumptions.

> It would be nice to add some usage example to board examples.

I'll try to do something.

> Dzięki za wkład  👍🏼

Żaden problem  👍🏼

glaeqen committed:
Stop storing *const () in a struct

Unimplements Sync and Send on anything that stores Dependencies,
usize solves this.

Pointer is used only for verification in `mcan`, so that is safe.

glaeqen committed:
Introduce feature-gates

glaeqen commented:
Fixed feature-gating for chips that have only single MCAN instance / have no CAN at all. Hopefully I didn't do it too restrictively.

glaeqen committed:
MCAN example

glaeqen committed:
Fix typo

glaeqen commented:
Tagging @sundbom-grepit

martinmortsell commented on hal/src/can.rs:
```diff
@@ -149,18 +150,22 @@ impl TxPin for Pin<PB2, PeripheralA> {
 }
 
+#[cfg(any(feature = "pins-100", feature = "pins-144"))]
 impl RxPin for Pin<PD28, PeripheralB> {
```
PD28 is only listed as CANRX1 on the 100 pin version.

martinmortsell replied:
Smells like an error in the datasheet to me, but if that's the case it should be verified that it works on the 144 pin version as well.

glaeqen replied:
No, I checked it and it does not work using that pin.

![image](https://user-images.githubusercontent.com/80407822/207364442-fd282762-6904-4f84-bd1c-8fdf20812b0d.png)

Good catch.

What we also noticed is that putting a pin in floating input does not disconnect the pin from peripherals. MCAN was still able to access RX from PC12 so I had to use external transceiver to verify it.

Interestingly enough, all peripheral inputs are always connected to the pin input. Does it mean that GPIO on x7x is unsound with current API?

![image](https://user-images.githubusercontent.com/80407822/207365962-7d5e7468-1028-4fd3-b30f-082919ecb811.png)

tmplt replied:
The API does not model this one-to-many input relationship, no.

glaeqen replied:
Such a poor HW design decision. Shame. Thanks for tuning in @tmplt

glaeqen committed:
Remove incorrect feature-gate

glaeqen committed:
Download `mcan-core` crate only for relevant MCUs

glaeqen committed:
Update CHANGELOG.md

martinmortsell commented on hal/src/can.rs:
```diff
@@ -0,0 +1,171 @@
+    /// - matrix.ccfg_can0.can0dmaba value cannot change
+    /// - matrix.ccfg_sysio.read().can1dmaba() value cannot change
+    pub unsafe fn new(
+        reg: Id::REG,
+        matrix: &crate::pac::MATRIX,
+        rx: Rx,
+        tx: Tx,
+        pck: &Pck<Pck5>,
+        hclk: &mut HostClock,
```
Does not match the format used by the other peripherals.
E.g. uart:
```rust
let mut uart = Uart::new_uart0(
    pac.UART0,
    (tx, rx),
    UartConfiguration::default(115_200.bps()),
    PeripheralClock::Other(&mut mck, &pck),
)
```
Pins in tuple, Configuration struct as input to `new()`.

glaeqen replied:
I can change pins, not sure what I can do differently regarding the rest.

tmplt replied:
Don't put down too much time refactoring for `PeripheralClock` if it isn't directly applicable. IIRC the wrapping struct is not fully sound.

martinmortsell commented on hal/src/can.rs:
```diff
@@ -0,0 +1,171 @@
+//! This module provides target specific integration with `mcan` crate.
```
I think the module should have more documentation than this.

glaeqen replied:
Added few sentences. In general, it's just bootstraping code to guarantee safe usage of an actual `mcan` crate.

glaeqen committed:
Extend docs a little

glaeqen committed:
Integrate pins into a tuple

glaeqen commented:
@michalfita @martinmortsell @tmplt is there anything missing? Should we merge it in?

glaeqen committed:
Switch to the crates.io reference for `mcan`

tmplt commented on boards/atsamv71_xult/memory.x:
```diff
@@ -1,7 +1,16 @@
+  .can (NOLOAD) :
+  {
+    *(.can .can.*);
+  } > CAN
```
What effect does this have? Region is untouched?

tmplt replied:
Thanks for the double post, github.

glaeqen replied:
It's like a dedicated `.bss` section, any symbol from `.can*` input sections should be mapped to `.can` output section in the CAN memory area. NOLOAD provides the expectation that this memory should not be initialized (usage through `MaybeUninit` is sound and `SharedMemory` is using it underneath). In general this is needed because CAN RAM has to be in a predictible memory region that matches platform's configuration. Eg. SAMD/E5x hardcodes it to 0x2000_XXXX and x7x is configurable by `CCFG_{CAN0, SYSIO}`. MCAN itself is unaware of this position, it is using 16-bit addressing internally.

tmplt commented on hal/src/can.rs:
```diff
@@ -0,0 +1,180 @@
+    /// # Safety
+    /// While [`Dependencies`] instance exists:
+    /// - matrix.ccfg_can0.can0dmaba value cannot change
+    /// - matrix.ccfg_sysio.read().can1dmaba() value cannot change
```
Thought for later, but `MATRIX` seems spaghetti enough that I wonder if it should be consumed upon clock configuration and unsafely accessed by the HAL when needed instead of passing a reference in multiple places.

Good on the `unsafe`. I believe we have another place where its state is assumed, but not appropriately `unsafe`d.

glaeqen replied:
That sounds like a sensible solution but maybe not up to this PR.

tmplt replied:
Indeed. Just putting the idea on paper.

tmplt commented:
I can't verify the example without hardware access, but otherwise lgtm.

tmplt commented:
No objections since last comment. Merging.
